### PR TITLE
feat: checkbox block

### PR DESCRIPTION
**Pull request by qinluhe**
### Feature Preview

- [x] Support convert both `-[] ` and `[] `to checkbox and unchecked, `-[x] ` and `[x] `to checkbox and checked.
- [x] Support cmd(macOS) or ctrl(windows or Linux) + enter to toggle the checkbox status.

image

---

#### PR Checklist

- [ ] My code adheres to the [AppFlowy Style Guide](https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/software-contributions/submitting-code/style-guides)
- [ ] I've listed at least one issue that this PR fixes in the description above.
- [ ] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [ ] All existing tests are passing.

### CODE REVIEW

**Comment by Xazin** [comment]
I haven't looked at the code, but for markdown behavior `[ ]` would be correct (notice the space in between the brackets).

More of an observation than anything.

**Comment by Xazin** [comment]
I see that the current pattern that is used is `/^(\[\])$/` 

With this pattern `/\[( |x)]/` you can test for the checked version `[x]` and the non-checked version `[ ]`, and I also think it should not matter if there is text preceding it, as you should be able to create checkboxes from existing text I presume?

But then I guess we want the checkbox to be the first thing on a line for it to be counted, so anchor it `/\[( |x)]/A`.

Just some notes, looks good though.

**Comment by qinluhe** [comment]
> I see that the current pattern that is used is `/^(\[\])$/`
> 
> With this pattern `/\[( |x)]/` you can test for the checked version `[x]` and the non-checked version `[ ]`, and I also think it should not matter if there is text preceding it, as you should be able to create checkboxes from existing text I presume?
> 
> But then I guess we want the checkbox to be the first thing on a line for it to be counted, so anchor it `/\[( |x)]/A`.
> 
> Just some notes, looks good though.
> 
> _(And of course, the space can be changed to `\s` for any whitespace so `/\[(\s|x)]/A`)_

I thought exactly the same as you did before, so I agree with your consideration.
But after I experimented with Notion's interaction, I still chose to refer to Notion

File: frontend/appflowy_tauri/src/appflowy_app/utils/document/slate/markdown.ts [hunk 1]
```diff
@@ -3,16 +3,28 @@ import { getBeforeRangeAt } from '$app/utils/document/slate/text';
...
+  const isHeadingMarkdown = /^(#{1,3})$/.test(flag);
+
+  return isHeadingMarkdown;
+}
+
+export function canHandleToCheckboxBlock(event: React.KeyboardEvent<HTMLDivElement>, editor: Editor) {
+  const flag = getMarkdownFlag(event, editor);
+  if (!flag) return false;
+
+  const isCheckboxMarkdown = /^(\[\])$/.test(flag);
```
Comment (LucasXu0) [inline frontend/appflowy_tauri/src/appflowy_app/utils/document/slate/markdown.ts:17]: FYI, we support convert both `-[]` and `[]` to uncheckbox, `-[x]` and `[x]` to checkbox.

**Comment by LucasXu0** [comment]
FYI, we support `cmd(macOS) or ctrl(windows or Linux) + enter` to toggle the checkbox status.

File: frontend/appflowy_tauri/src/appflowy_app/components/document/CheckboxBlock/index.tsx [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  node: NestedBlock<BlockType.CheckboxBlock>;
+  childIds?: string[];
+}) {
+  const { id, data } = node;
+  const { toggleCheckbox } = useCheckboxBlock(id, node.data);
+
+  const checked = !!data.checked;
+
+  return (
+    <>
```
[the remark this answers is not in the document]
Comment (Xazin) [reply frontend/appflowy_tauri/src/appflowy_app/components/document/CheckboxBlock/index.tsx:22]: This is often done in React to have multiple children, you can view it as a raw `div` tag basically.

In this case it's used like this:
```
<>
  <div>..</div>
  <NodeChildren />
</>
```

File: frontend/appflowy_tauri/src/appflowy_app/stores/reducers/document/async-actions/blocks/text/index.ts [hunk 3]
```diff
@@ -1,39 +1,32 @@
 import { createAsyncThunk } from '@reduxjs/toolkit';
 import { DocumentController } from '$app/stores/effects/document/document_controller';
-import { DocumentState } from '$app/interfaces/document';
-import { newTextBlock } from '$app/utils/document/blocks/text';
-import { setCursorBeforeThunk } from '$app_reducers/document/async-actions/cursor';
+import { BlockType, DocumentState } from '$app/interfaces/document';
+import { turnToBlockThunk } from '$app_reducers/document/async-actions/trun_to';
```
[the remark this answers is not in the document]
Comment (Xazin) [reply frontend/appflowy_tauri/src/appflowy_app/stores/reducers/document/async-actions/blocks/text/index.ts:4]: What is `Thunk`

[the remark this answers is not in the document]
Comment (qinluhe) [reply frontend/appflowy_tauri/src/appflowy_app/stores/reducers/document/async-actions/blocks/text/index.ts:4]: > 

https://redux-toolkit.js.org/api/createAsyncThunk

**Comment by LucasXu0** [comment]
LGTM.

**Pull request merged**